### PR TITLE
[SNOW-1246906] Fetch available packages list from Snowflake instead of directly from Anaconda

### LINKED ISSUE #908: SNOW-1246906: `snow build` snowflake/core dependency bundling broken
### SnowCLI version

2.1.0

### Python version

3.11.7

### Platform

macOS-10.16-x86_64-i386-64bit

### What happened

I am deploying a snowpark project that uses snowflake and snowflake.core packages. 
If I add `snowflake` and `snowflake.core` to the `requirements.txt`, and run `snow build`, the `snowflake` and `snowflake.core` dependencies are added to a `requirements.other.txt` instead of the `requirements.snowflake.txt`, so running `snow deploy` will not include the snowflake packages in the `packages` list of the procedure or function being deployed, so the deployment fails.

If I manually copy the `snowflake` and `snowflake.core` into the `requirements.snowflake.txt` file, after building and before deploying, it works as expected. This seems to just be a matter of adding snowflake (and related) dependencies to the correct `requirements.txt` file.

### Console output

```Shell
build succeeds, deploy fails do to missing `snowflake.core` package.
```

### How to reproduce

1. create a new directory
2. navigate into new directory
3. initialize a new snowpark project
4. update procedure/function to use snowflake.core, like retrieving a TaskResource from a db/schema.
5. add `snowflake` and/or `snowflake.core` to the `requirements.txt`
5. build the project using `snow build`

you will see the packages are add to `requirements.other.txt` instead of `requirements.snowflake.txt`

### PR DESCRIPTION
**Pull request by sfc-gh-pjob**
### Pre-review checklist
   * [x] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [x] I've added or updated automated unit tests to verify correctness of my new code.
   * [x] I've added or updated integration tests to verify correctness of my new code.
   * [x] I've confirmed that my changes are working by executing CLI's commands manually.
   * [x] I've confirmed that my changes are up-to-date with the target branch.
   * [x] I've described my changes in the release notes.
   * [x] I've described my changes in the section below.

### Changes description
* Fetching available packages list from Snowflake instead of directly from Anaconda with fallback to the old method (for backward compatibility).
  As the new method requires a connection to Snowflake, it adds connection options to the following commands:
  * `snow snowpark build`
  * `snow snowpark package lookup`
  * `snow snowpark package create`
* Fixes #908

### CODE REVIEW

File: tests/test_utils.py [hunk 1]
```diff
@@ -118,35 +117,6 @@ def test_parse_requirements(correct_requirements_txt: str):
     assert result[2].specs == [("==", "1.0.0")]
 
 
-@mock.patch("requests.get")
-def test_filter_anaconda_packages(mock_get):
```
Comment (sfc-gh-pjob) [inline tests/test_utils.py:122]: This test was duplicated in `test_anaconda.py` (now in `test_packages_available_in_snowflake`)

File: tests/snowpark/test_packages_available_in_snowflake.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        "1.0.0",
+        "0.1.1",
+    ]
+    assert PACKAGES_AVAILABLE_IN_SNOWFLAKE.package_versions(
+        Requirement.parse("jpeg")
+    ) == ["9e", "9d", "9b"]
+    assert (
+        PACKAGES_AVAILABLE_IN_SNOWFLAKE.package_versions(Requirement.parse("weird-pkg"))
+        == []
+    )
```
Comment (sfc-gh-astus) [inline tests/snowpark/test_packages_available_in_snowflake.py:91]: Why not pytest.mark.parametrize?

File: src/snowflake/cli/plugins/snowpark/package/packages_in_snowflake.py [hunk 3]
```diff
@@ -171,3 +148,70 @@ def write_requirements_file_in_snowflake_format(
 
         if formatted_requirements:
             file_path.write_text("\n".join(formatted_requirements))
+
+
+class PackagesAvailableInSnowflakeManager(SqlExecutionMixin):
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/plugins/snowpark/package/packages_in_snowflake.py:153]:
```suggestion
class AnacondaPackagesManager(SqlExecutionMixin):
```
WDYT? We already have non-anaconda packages (from package create or snowpark build) in Snowflake, I feel that we should not drop Anaconda from name

File: src/snowflake/cli/plugins/snowpark/package/packages_in_snowflake.py [hunk 4]
```diff
@@ -171,3 +148,70 @@ def write_requirements_file_in_snowflake_format(
...
+
+    def _query_snowflake_for_available_packages(self) -> dict[str, AvailablePackage]:
+        cursor = self._execute_query(
+            "select package_name, version from information_schema.packages where language = 'python'"
+        )
+        if cursor.rowcount is None or cursor.rowcount == 0:
+            raise SnowflakeSQLExecutionError()
+        packages: dict[str, AvailablePackage] = {}
+        for row in cursor:
+            if not (package_name := row[0]):
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/plugins/snowpark/package/packages_in_snowflake.py:185]: Have you considered executing the query with `cursor_class=DictCursor`? In this way we don't have to operate on indexes

File: tests/snowpark/test_packages_available_in_snowflake.py (hunk 2 above)
Comment (sfc-gh-pjob) [reply tests/snowpark/test_packages_available_in_snowflake.py:91]: This is the old test (from `test_anaconda.py`).

File: src/snowflake/cli/plugins/snowpark/package/packages_in_snowflake.py (hunk 3 above)
Comment (sfc-gh-pjob) [reply src/snowflake/cli/plugins/snowpark/package/packages_in_snowflake.py]: I used your suggestion

File: src/snowflake/cli/plugins/snowpark/package/packages_in_snowflake.py (hunk 4 above)
Comment (sfc-gh-pjob) [reply src/snowflake/cli/plugins/snowpark/package/packages_in_snowflake.py]: Ok, I used DictCursor

**Review by sfc-gh-astus — APPROVED** [review]

**Review by sfc-gh-turbaszek — APPROVED** [review]

**Pull request merged**